python/lsst/alert/packet/schema.py

```diff
@@ -0,0 +1,203 @@
+    """
+    def __init__(self, schema_definition):
+        self.definition = schema_definition
+
+    def serialize(self, record):
+        """Create an Avro representation of data following this schema.
+
+        Parameters
+        ----------
+        record : `dict`
+            The data to be serialized to Avro.
+
+        Returns
+        -------
+        avro_data : `bytes`
+            An Avro serialization of the input data.
+        """
+        bytes_io = io.BytesIO()
+        fastavro.schemaless_writer(bytes_io, self.definition, record)
+        return bytes_io.getvalue()
+
+    def deserialize(self, record):
+        """Deserialize an Avro packet folowing this schema.
+
+        Parameters
+        ----------
+        record : `bytes`
+            The data to be deserialized.
+
+        Returns
+        -------
+        alert_data : `dict`
+            Deserialized packet contents.
+        """
+        bytes_io = io.BytesIO(record)
+        message = fastavro.schemaless_reader(bytes_io, self.definition)
+        return message
+
+    def validate(self, record):
+        """Validate packet contents against this schema.
+
+        Parameters
+        ----------
+        record : `dict`
+            The data to be checked for schema compliance.
+
+        Returns
+        -------
+        valid : `bool`
+            Whether or not the data complies with the schema.
+        """
+        return fastavro.validate(record, self.definition)
+
+    def store_alerts(self, fp, records):
+        """Store alert packets to the given I/O stream.
+
+        Parameters
+        ----------
+        fp : derivative of `IOBase`
+            I/O stream to which data will be written.
+        records : iterable of `dict`
+            Alert records to be stored.
+        """
+        fastavro.writer(fp, self.definition, records)
+
+    def retrieve_alerts(self, fp):
+        """Read alert packets from the given stream.
+
+        Parameters
+        ----------
+        fp : derivative of `IOBase`
+            I/O stream from which data will be read.
+        schema : `list`, optional
+            A schema describing the contents of the Avro packets. If not
+            provided, the schema used when writing the alert steram will be
+            used.
+
+        Returns
+        -------
+        schema : `lsst.alert.Schema`
+            The schema with which alerts were written (which may be different
+            from this schema being used for deserialization).
+        records : iterable of `dict`
+            Alert records.
+        """
+        from .io import retrieve_alerts
+        return retrieve_alerts(fp, reader_schema=self)
+
+    def __eq__(self, other):
+        """Compare schemata for equality.
+
+        Schemata are regarded as equal if their fully-resolved definitions are
+        the same.
+        """
+        return self.definition == other.definition
+
+    @staticmethod
+    def resolve(schema):
```

Comment:
Interesting idea —  I hadn't spotted that fastavro issue. Reading it gave me some ideas to improve my implementation, which I've done, and I've added a test.

Figuring out how best to upstream this will take some work (they'll have their own testing standards, and I'll need to figure out how to slot it into their codebase, and I've probably made some naïve mistakes and will have to iterate with them on the implementation. I'll make a ticket to do this, but I won't have time to work on it immediately.

This is [DM-19264](https://jira.lsstcorp.org/browse/DM-19264).